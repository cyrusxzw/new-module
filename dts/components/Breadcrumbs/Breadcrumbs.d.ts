import { FC } from 'react';
import type { BreadcrumbsProps } from './Breadcrumbs.types';
declare const Breadcrumbs: FC<BreadcrumbsProps>;
export { Breadcrumbs };
