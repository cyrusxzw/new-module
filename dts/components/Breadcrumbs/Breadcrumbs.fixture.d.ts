import type { BreadcrumbsProps } from './Breadcrumbs.types';
declare const BreadcrumbsFixture: BreadcrumbsProps;
export { BreadcrumbsFixture };
