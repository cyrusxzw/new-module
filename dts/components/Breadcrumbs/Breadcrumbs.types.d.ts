import type { Themes } from '~/types';
declare type BreadcrumbItem = {
  label: string;
  id: string;
  url: string;
  title: string;
};
declare type BreadcrumbsProps = {
  className?: string;
  items?: BreadcrumbItem[];
  theme?: Themes;
};
export type { BreadcrumbsProps };
