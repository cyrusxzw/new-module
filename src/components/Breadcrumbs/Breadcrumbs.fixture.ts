import type { BreadcrumbsProps } from './Breadcrumbs.types';

const BreadcrumbsFixture: BreadcrumbsProps = {
  className: 'breadcrumbs',
  items: [
    {
      label: 'Suspendisse quis elit',
      id: 'suspendisse-quis-elit',
      url: '/',
      title: 'Breadcrumb title 1',
      dataTestRef: 'test_id_1',
    },
    {
      label: 'Maecenas luctus',
      id: 'maecenas-luctus',
      url: '/',
      title: 'Breadcrumb title 2',
      dataTestRef: 'test_id_2',
    },
  ],
  theme: 'dark',
};

export { BreadcrumbsFixture };
