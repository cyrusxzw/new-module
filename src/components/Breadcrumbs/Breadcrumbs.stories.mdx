import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import { action } from '@storybook/addon-actions';
import * as knobs from '@storybook/addon-knobs';
import { Breadcrumbs } from './Breadcrumbs';
import { BreadcrumbsFixture } from './Breadcrumbs.fixture';

<Meta title="Components/Breadcrumbs" component={Breadcrumbs} />

# Breadcrumbs

<Canvas>
  <Story name="Base component">
    <Breadcrumbs
      items={knobs.object('items', BreadcrumbsFixture.items, 'Copy')}
      onHyperlinkClick={action('clicked')}
      theme={knobs.select(
        'theme',
        ['dark', 'light'],
        BreadcrumbsFixture.theme,
        'Presentation',
      )}
    />
  </Story>
</Canvas>

## API

<ArgsTable of={Breadcrumbs} />
