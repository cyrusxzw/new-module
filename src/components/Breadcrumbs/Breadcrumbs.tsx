import React, { FC } from 'react';
import cx from 'classnames';
import { isObjectPopulatedArray } from '~/utils/objects';
import { Hyperlink } from '~/components/Hyperlink';
import type { BreadcrumbsProps } from './Breadcrumbs.types';
import styles from './Breadcrumbs.module.css';

const Breadcrumbs: FC<BreadcrumbsProps> = ({
  className,
  items,
  theme = 'dark',
}) => {
  if (!items || !isObjectPopulatedArray(items)) {
    return null;
  }

  const classSet = cx(styles.base, styles[theme], className);

  return (
    <nav className={classSet} data-testid="data-testid-Breadcrumbs">
      <ul className={styles.list}>
        {items.map(item => (
          <li className={styles.item} key={item.id}>
            <Hyperlink
              className={styles.link}
              id={item.id}
              theme={theme}
              title={item.title}
              url={item.url}
            >
              {item.label}
            </Hyperlink>
          </li>
        ))}
      </ul>
    </nav>
  );
};

export { Breadcrumbs };
