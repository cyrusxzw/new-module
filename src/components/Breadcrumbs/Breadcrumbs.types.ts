import type { Themes } from '~/types';

type BreadcrumbItem = {
  label: string;
  id: string;
  url: string;
  title: string;
};

type BreadcrumbsProps = {
  className?: string;
  items?: BreadcrumbItem[];
  onHyperlinkClick?: (item: BreadcrumbItem) => void;
  theme?: Themes;
};

export type { BreadcrumbsProps };
