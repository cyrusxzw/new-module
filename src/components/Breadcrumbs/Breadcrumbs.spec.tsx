import React from 'react';
import { render, screen } from '@testing-library/react';
import { Breadcrumbs } from './Breadcrumbs';
import { BreadcrumbsFixture } from './Breadcrumbs.fixture';

describe('<Breadcrumbs />', () => {
  it('should render base component correctly', () => {
    const { container } = render(
      <Breadcrumbs
        className={BreadcrumbsFixture.className}
        items={BreadcrumbsFixture.items}
        theme={BreadcrumbsFixture.theme}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should return `null` if the length of the `items` prop is 0', () => {
    render(<Breadcrumbs items={[]} />);

    const breadcrumbs = screen.queryByTestId(/data-testid-Breadcrumbs/i);

    expect(breadcrumbs).not.toBeInTheDocument();
  });

  it('should render the corrct number of items', async () => {
    render(<Breadcrumbs items={BreadcrumbsFixture.items} />);

    const listItems = screen.getAllByRole('listitem');

    expect(listItems).toHaveLength(2);
  });
});
