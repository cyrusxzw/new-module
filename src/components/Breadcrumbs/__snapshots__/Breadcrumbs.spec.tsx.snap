// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumbs /> should render base component correctly 1`] = `
<div>
  <nav
    class="base dark breadcrumbs"
    data-testid="data-testid-Breadcrumbs"
  >
    <ul
      class="list"
    >
      <li
        class="item"
      >
        <a
          class="base left dark link"
          data-test-ref="test_id_1"
          href="/"
          target="_self"
          title="Breadcrumb title 1"
        >
          Suspendisse quis elit
        </a>
      </li>
      <li
        class="item"
      >
        <a
          class="base left dark link"
          data-test-ref="test_id_2"
          href="/"
          target="_self"
          title="Breadcrumb title 2"
        >
          Maecenas luctus
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
